src/serializer.js

```diff
@@ -91,8 +91,86 @@ class FullPacketParser extends Transform {
   }
 }
 
+function recursiveGet (target, path, receiver) {
+  return path.reduce((object, key) => Reflect.get(object, key, receiver), target)
+}
+
+class LazyPacketParser extends Transform {
+  constructor (proto, mainType, shallowType, noErrorLogging = false) {
+    super({ readableObjectMode: true })
+    this.proto = proto
+    this.mainType = mainType
+    this.shallowType = shallowType
+    this.noErrorLogging = noErrorLogging
+  }
+
+  parsePacketBuffer (buffer) {
+    let shallowPacketData
+    try {
+      shallowPacketData = this.proto.parsePacketBuffer(this.shallowType, buffer).data
+    } catch (e) {
+      return this.proto.parsePacketBuffer(this.mainType, buffer)
+    }
```

Comment:
Can't understand why this is here, if ProtoDef fails to parse a buffer using shallow type (ex. incomplete/corrupted buffer), then why it should parse again but with main type?

Reply:
Ah this was so the user didn't have to make sure the shallow type was compatible with all packets that might get parsed. If shallow parsing failed, it would just fall back to parsing with the main type. But we can take this part out and throw an error if shallow parsing fails.